service/orchestrator/catalogs.go

```diff
@@ -0,0 +1,171 @@
+	if req.CatalogId == "" {
+		return nil, status.Errorf(codes.InvalidArgument, "catalog id is empty")
+	}
+
+	if req.Catalog == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "catalog is empty")
+	}
+
+	res = req.Catalog
+	res.Id = req.CatalogId
+
+	err = svc.storage.Update(res, "id = ?", res.Id)
+
+	if err != nil && errors.Is(err, persistence.ErrRecordNotFound) {
+		return nil, status.Error(codes.NotFound, "catalog not found")
+	} else if err != nil {
+		return nil, status.Errorf(codes.Internal, "database error: %v", err)
+	}
+	return
+}
+
+// RemoveCatalog implements method for removing a catalog
+func (svc *Service) RemoveCatalog(_ context.Context, req *orchestrator.RemoveCatalogRequest) (response *emptypb.Empty, err error) {
+	if req.CatalogId == "" {
+		return nil, status.Errorf(codes.InvalidArgument, "catalog id is empty")
+	}
+
+	err = svc.storage.Delete(&orchestrator.Catalog{}, "Id = ?", req.CatalogId)
+	if errors.Is(err, persistence.ErrRecordNotFound) {
+		return nil, status.Errorf(codes.NotFound, "catalog not found")
+	} else if err != nil {
+		return nil, status.Errorf(codes.Internal, "database error: %v", err)
+	}
+
+	return &emptypb.Empty{}, nil
+}
+
+func (srv *Service) GetCategory(ctx context.Context, req *orchestrator.GetCategoryRequest) (res *orchestrator.Category, err error) {
+	res = new(orchestrator.Category)
+	err = srv.storage.Get(&res, gorm.WithPreload("Controls", "parent_control_short_name IS NULL"), "name = ? AND catalog_id = ?", req.CategoryName, req.CatalogId)
+	if errors.Is(err, persistence.ErrRecordNotFound) {
+		return nil, status.Errorf(codes.NotFound, "category not found")
+	} else if err != nil {
+		return nil, status.Errorf(codes.Internal, "database error: %v", err)
+	}
+
+	return res, nil
+}
+
+func (srv *Service) GetControl(ctx context.Context, req *orchestrator.GetControlRequest) (res *orchestrator.Control, err error) {
+	res = new(orchestrator.Control)
+	err = srv.storage.Get(&res, "short_name = ? AND category_name = ? AND category_catalog_id = ?", req.ControlShortName, req.CategoryName, req.CatalogId)
+	if errors.Is(err, persistence.ErrRecordNotFound) {
+		return nil, status.Errorf(codes.NotFound, "control not found")
+	} else if err != nil {
+		return nil, status.Errorf(codes.Internal, "database error: %v", err)
+	}
+
+	return res, nil
+}
+
+func (srv *Service) ListControls(ctx context.Context, req *orchestrator.ListControlsRequest) (res *orchestrator.ListControlsResponse, err error) {
+	// Validate the request
+	if err = api.ValidateListRequest[*orchestrator.Control](req); err != nil {
+		err = fmt.Errorf("invalid request: %w", err)
+		log.Error(err)
+		err = status.Errorf(codes.InvalidArgument, "%v", err)
+		return
+	}
+
+	res = new(orchestrator.ListControlsResponse)
+
+	// If the category name is set (additional binding), forward it as a condition to the pagination method
+	if req.CategoryName != "" {
```

Comment:
One probably should specify the category name AND the catalog ID and if both are set, then follow the `if`, since category names are only unique within a catalog.